### PR TITLE
VUFIND-1342 Modernize browse indexer

### PR DESCRIPTION
**Pull request by marktriggs**
This PR addresses the issues described here:

https://openlibraryfoundation.atlassian.net/browse/VUFIND-1342

I've finished my first pass across everything, and thought I'd get a PR up to invite feedback.  The story so far:

  * The unqualified classes used for indexing are now in the `org.vufind.solr.indexing` package

  * The word "leech" has been replaced with "FieldIterator" in the places it appeared

  * All FieldIterators are autoclosable, and we take advantage of that where possible

  * Cleaned up imports (expanded wildcards and removed unused imports)

 To preserve compatibility with existing usage, there are wrapper versions of `CreateBrowseSQLite` and `PrintBrowseHeadings` under a `compat` directory that log deprecation warnings and then pass through to the new versions.  There's also code to detect usage of old environment variables and system properties (BIBLEECH) and old versions of class names.  All of this code can probably be thrown out in a future release, once people have had time to update their scripts.

Besides that, tests are all still passing, and I have a corresponding PR against VuFind to update the `index-alphabetic-browse` scripts to use the new names.  I'll link to this PR from there once I send it.

TODO
- [x] When merging, open JIRA ticket to track task to remove deprecated code in next major release.

### CODE REVIEW

File: README.md
Comment (demiankatz) [inline README.md]: We could specifically say "versions included in VuFind® 9,x and earlier" -- since the goal is to get this PR merged in time for release 10.0.

File: src/main/java/org/vufind/solr/indexing/PrintBrowseHeadings.java
Comment (demiankatz) [inline src/main/java/org/vufind/solr/indexing/PrintBrowseHeadings.java:30]: It's odd that Git isn't recognizing this as a file rename, even though all the others got picked up correctly! Oh well.

**Review by demiankatz — CHANGES_REQUESTED** [review]
Thanks, @marktriggs, this looks great! The test suite is passing, so I've gone ahead and pushed the binaries into https://github.com/vufind-org/vufind/pull/3253 so it is ready to merge when this is finished. (And of course I'll update them if other feedback leads to changes in the meantime... but since I had to put them there to test, I figured I might as well commit in case it saves us time in future).

See below for one tiny documentation suggestion, and an observation of no great importance.

File: README.md
Comment (marktriggs) [reply README.md]: I've just added a note for this, thanks!

File: src/main/java/org/vufind/solr/indexing/PrintBrowseHeadings.java
Comment (marktriggs) [reply src/main/java/org/vufind/solr/indexing/PrintBrowseHeadings.java:30]: Yeah.  By default, git requires files at different paths to be 50% similar to detect renames, and `PrintBrowseHeadings` bore the brunt of the changes associated with the class renames and autoclose changes, so it's tipped over that threshold.

On the command line I can get it to pick up the rename by reducing that threshold to 20%:

```git diff --stat -M0.2 origin/dev VUFIND-1342```

but that doesn't help the web UI much :)

**Review by demiankatz — APPROVED** [review]
Thanks, @marktriggs! I'm approving this, but I'll hold off on merging it for a few days in case @todolson or others have any additional comments.

**Review by todolson — APPROVED** [review]
I have not fully (re?)digested `PrintBrowseHeadings.java`, but so far I do think this all looks good. I see no reason to stand in the way.

Thank you, @marktriggs, for taking on this overdue modernization!

File: src/main/java/org/vufind/solr/indexing/PrintBrowseHeadings.java
Comment (todolson) [inline src/main/java/org/vufind/solr/indexing/PrintBrowseHeadings.java]: Late comment: would be good to change this parameter description, since the Leech classes have all been renamed. Maybe something like:
"Iterator for pulling in headings from Solr matches"

File: src/main/java/org/vufind/solr/indexing/PrintBrowseHeadings.java
Comment (todolson) [inline src/main/java/org/vufind/solr/indexing/PrintBrowseHeadings.java]: This is probably out of scope for the current effort, but this makes me wonder if `SolrFieldIterator` should implement the `Iterator` interface sometime in the future. It already has a `next()` method, so it's mostly there.

**Review by todolson — APPROVED** [review]
A couple items that I didn't catch on Friday. These comments are not meant to derail a commit.

**Review by demiankatz — CHANGES_REQUESTED** [review]
I agree with @todolson that it's probably a good idea to change the comment(s) that use the word "Leech" for consistency, but more substantive changes can be addressed separately. Do you mind making the edit, @marktriggs, or would you like me to take care of it for you?

**Comment by marktriggs** [comment]
> I agree with @todolson that it's probably a good idea to change the comment(s) that use the word "Leech" for consistency, but more substantive changes can be addressed separately. Do you mind making the edit, @marktriggs, or would you like me to take care of it for you?

I can pick that up since I'm still set up for testing, thanks for the review @todolson.  I'll look at adding the Iterator interface too; might as well :)

**Comment by demiankatz** [comment]
Thanks! I'll wait for this to get finished before I take on #49, though it looks like that's going to be an easy one to merge. :-)

**Comment by marktriggs** [comment]
@demiankatz @todolson I think that's ready to go now.  I ended up implementing both `Iterator` and `Iterable`, since the latter is an easy addition and it neatens up the loop over headings a little.

**Review by demiankatz — APPROVED** [review]
This looks good to me now, and all tests still pass.

**Pull request merged**
